### PR TITLE
Add prepared statements support for `Mysql2Adapter`

### PR DESCRIPTION
**Pull request by kamipo**
Reopen #22415.

Currently mysql2 adapter does not support prepared statements. And mysql adapter has been removed. This means that Active Record does not support prepared statements for MySQL now.

We need to support prepared statements for MySQL. But mysql2 gem still have GC issue in prepared statements. https://github.com/brianmario/mysql2/issues/694

I made default to `@prepared_statements = false` in `Mysql2Adapter` for does not affect default behavior. I think it is still safety.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb [hunk 1]
```diff
@@ -126,6 +141,10 @@ def supports_datetime_with_precision?
         version >= '5.6.4'
       end
 
+      def supports_json?
```
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: This seems unrelated to prepared statement support

**Comment by kamipo** [comment]
Extracted unrelated change to #23469, thanks!
r? @sgrif

**Comment by sgrif** [comment]
This will need to handle nodes which we consider to be unpreparable, similar to what we do in PG and SQLite. We can probably just move that logic up to the abstract adapter now.

**Comment by kamipo** [comment]
Do you means #23515?

**Comment by sodabrew** [comment]
I've just pushed mysql2 0.4.3 with important prepared statement fixes from @kamipo!

**Comment by jeremy** [comment]
Rebase time 😊

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb [hunk 2]
```diff
@@ -128,28 +141,61 @@ def execute(sql, name = nil)
...
-        execute to_sql(sql, binds), name
-        @connection.affected_rows
+      def exec_query(sql, name = 'SQL', binds = [], prepare: false) # :nodoc:
+        if without_prepared_statement?(binds)
+          execute_and_free(sql, name) do |result|
+            ActiveRecord::Result.new(result.fields, result.to_a) if result
+          end
+        else
+          exec_stmt_and_free(sql, name, binds, cache_stmt: prepare) do |_, result|
+            ActiveRecord::Result.new(result.fields, result.to_a) if result
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: In mysql2 0.4.3, `result.fields` causes segv when num of rows is zero.
This issue was fixed on master by https://github.com/brianmario/mysql2/pull/741.
@sodabrew Could you release mysql2 0.4.4?

**Comment by sodabrew** [comment]
I've just posted mysql2 0.4.4 with additional prepared statement fixes from @kamipo!

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb (hunk 2 above)
Comment (sodabrew) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: Done! mysql2 0.4.4 is now on Rubygems.org.

**Comment by kamipo** [comment]
Thanks @sodabrew !! 😍

**Comment by kamipo** [comment]
Rebased on master!

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb [hunk 3]
```diff
@@ -178,6 +195,14 @@ def disable_referential_integrity #:nodoc:
         end
       end
 
+      # CONNECTION MANAGEMENT ====================================
+
+      # Clears the prepared statements cache.
+      def clear_cache!
+        reload_type_map
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: Should we call `super` here?

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb [hunk 4]
```diff
@@ -35,9 +36,11 @@ module ConnectionAdapters
     class Mysql2Adapter < AbstractMysqlAdapter
       ADAPTER_NAME = 'Mysql2'.freeze
 
+      include MySQL::DatabaseStatements
+
       def initialize(connection, logger, connection_options, config)
         super
-        @prepared_statements = false
+        @prepared_statements = false unless config.key?(:prepared_statements)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: Should we use the value of the `prepared_statements` config here like we do in the other adapters?

File: activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+          end
+          @connection.next_result while @connection.more_results?
+          result
+        end
+
+        # Returns a record hash with the column names as keys and column values
+        # as values.
+        def select_one(arel, name = nil, binds = [])
+          arel, binds = binds_from_relation(arel, binds)
+          @connection.query_options.merge!(as: :hash)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb]: What this does?

**Comment by jeremy** [comment]
Changing to 5.0 since prepared statements are [disabled by default](https://github.com/rails/rails/pull/23461/files#diff-e7dead35794529e0cc5d0b2d788f8235R43).

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb [hunk 6]
```diff
@@ -191,19 +216,11 @@ def explain(arel, binds = [])
...
-
       # Executes the SQL statement in the context of this connection.
       def execute(sql, name = nil)
         log(sql, name) { @connection.query(sql) }
       end
 
-      # Mysql2Adapter doesn't have to free a result after using it, but we use this method
-      # to write stuff in an abstract way without concerning ourselves about whether it
-      # needs to be explicitly freed or not.
       def execute_and_free(sql, name = nil) # :nodoc:
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: We can remove this method here in a different PR. This abstract class can be removed too since we only have one concrete class.

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb (hunk 3 above)
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: PG and SQLite3 adapters does not call `super`.
https://github.com/rails/rails/blob/v5.0.0.beta3/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L230-L232
https://github.com/rails/rails/blob/v5.0.0.beta3/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L149-L151

https://github.com/rails/rails/blob/v5.0.0.beta3/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L356-L358

Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: :+1:

File: activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb (hunk 5 above)
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb]: `stmt.execute` calls `result.each` in the internal.
https://github.com/brianmario/mysql2/blob/0.4.4/ext/mysql2/statement.c#L384
Need to set `query_options` before `stmt.execute`.

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb (hunk 4 above)
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: This is for prepared_statements disabled by default (keep previous default).
Should we make to enabled by default?

File: activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb (hunk 5 above)
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb]: But how returning a hash instead of array in the result would be affected by the `result.each` call?

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb (hunk 4 above)
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: We should respect it. If users ask to enable it, it should be enabled, to disable it, it should be disabled. This implementation while keeping the default is enabling if you do `prepared_statements: false` in your config.

File: activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb (hunk 5 above)
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb]: `result.each` (`rb_mysql_result_each`) passes `@query_options` to `fetch_row_func` (`rb_mysql_result_fetch_row_stmt ` or `rb_mysql_result_fetch_row `).
https://github.com/brianmario/mysql2/blob/0.4.4/ext/mysql2/result.c#L871-L883
https://github.com/brianmario/mysql2/blob/0.4.4/ext/mysql2/result.c#L929-L944

`fetch_row_func` handles `args->asArray`.
https://github.com/brianmario/mysql2/blob/0.4.4/ext/mysql2/result.c#L510-L514
https://github.com/brianmario/mysql2/blob/0.4.4/ext/mysql2/result.c#L725-L729

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb (hunk 4 above)
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: This implementation is:

```ruby
if config.key?(:prepared_statements)
  # @prepared_statements is configured in `super`
  super
else
  # keep the previous default (disabled) when `:prepared_statements` is not specified
  @prepared_statements = false
end

File: activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb (hunk 5 above)
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb]: I see now. Thanks for the explanation!

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb (hunk 4 above)
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: Oh. I missed that change that moved AbstractAdapter to configure prepared_statements 👍

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb [hunk 7]
```diff
@@ -1,6 +1,7 @@
 require 'active_record/connection_adapters/abstract_mysql_adapter'
+require 'active_record/connection_adapters/mysql/database_statements'
 
-gem 'mysql2', '>= 0.3.18', '< 0.5'
+gem 'mysql2', '~> 0.4.4'
```
Comment (jeremy) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: Do we need to bump mysql2 minimum version? 0.4.x is only required if `prepared_statements: true`.

Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: Unfortunately, mysql2 0.4.3 was broken without prepared statements.
I would like to skip this version.

**Comment by sodabrew** [comment]
Hurray!
